--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,4 +1,20 @@
-cmake_minimum_required (VERSION 3.16.4)
+#
+# Copyright (c) 2014 Brent Cook
+#
+# Permission to use, copy, modify, and distribute this software for any
+# purpose with or without fee is hereby granted, provided that the above
+# copyright notice and this permission notice appear in all copies.
+#
+# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+cmake_minimum_required (VERSION 3.17.5)
+
 if(MSVC)
 	cmake_policy(SET CMP0091 NEW)
 endif()
@@ -371,18 +387,23 @@ if(APPLE AND (NOT CMAKE_OSX_ARCHITECTURES STREQUAL ""))
 	set(CMAKE_SYSTEM_PROCESSOR "${CMAKE_OSX_ARCHITECTURES}")
 endif()
+# CMAKE_SYSTEM_PROCESSOR is not consistently set to the target architecture.
+# https://cmake.org/cmake/help/latest/variable/CMAKE_SYSTEM_PROCESSOR.html
+if (WIN32 AND (NOT CMAKE_GENERATOR_PLATFORM STREQUAL ""))
+	message("CMAKE_SYSTEM_PROCESSOR: ${CMAKE_SYSTEM_PROCESSOR}")
+	message("CMAKE_GENERATOR_PLATFORM: ${CMAKE_GENERATOR_PLATFORM}")
+	set(CMAKE_SYSTEM_PROCESSOR "${CMAKE_GENERATOR_PLATFORM}")
+endif()
+
 if(CMAKE_SYSTEM_PROCESSOR MATCHES "(aarch64|arm64|ARM64)")
 	set(HOST_AARCH64 true)
-	if(WIN32)
-		set(ENABLE_ASM false)
-	endif()
 elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "arm")
 	set(HOST_ARM true)
 elseif(CMAKE_SYSTEM_NAME STREQUAL "SunOS" AND CMAKE_SYSTEM_PROCESSOR MATCHES "i386")
 	set(HOST_X86_64 true)
-elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "(x86_64|amd64|AMD64)")
+elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "(x86_64|x64|amd64|AMD64)")
 	set(HOST_X86_64 true)
-elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "(i[3-6]86|[xX]86)")
+elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "(i[3-6]86|[xX]86|Win32)")
 	set(ENABLE_ASM false)
 	set(HOST_I386 true)
 elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "loongarch64")
 
@@ -407,8 +428,6 @@ if(ENABLE_ASM)
 	if(CMAKE_C_COMPILER_ABI STREQUAL "ELF")
 		if(HOST_X86_64)
 			set(HOST_ASM_ELF_X86_64 true)
-		elseif(HOST_ARM)
-			set(HOST_ASM_ELF_ARMV4 true)
 		elseif(CMAKE_SYSTEM_NAME STREQUAL "SunOS" AND HOST_I386)
 			set(HOST_ASM_ELF_X86_64 true)
 		endif()
@@ -532,8 +551,8 @@ if(ENABLE_LIBRESSL_INSTALL)
 		# Create pkgconfig files.
 		set(prefix      ${CMAKE_INSTALL_PREFIX})
 		set(exec_prefix \${prefix})
-		set(libdir      \${exec_prefix}/${CMAKE_INSTALL_LIBDIR})
-		set(includedir  \${prefix}/${CMAKE_INSTALL_INCLUDEDIR})
+    set(libdir      \${exec_prefix}/lib)
+    set(includedir  \${prefix}/include)
 		if(PLATFORM_LIBS)
 			string(REGEX REPLACE ";" " -l" PLATFORM_LDADD ";${PLATFORM_LIBS}")
 		endif()
